docs/rfcs/rfc04.md

```diff
@@ -0,0 +1,551 @@
+
+```jsonnet
+// app/app.jsonnet
+(import 'k8s.libsonnet').RootComponent {
+  deployment: $.k8s.apps.v1.Deployment {
+    // ..
+  spec+: {
+    template+: {
+      spec+: {
+        containers_+: {
+          foo: {
+            image: $.inventory.foo.image,
+            // ...         
+          },
+        },
+      }
+    },
+  },
+}
+```
+
+Each "phase" can point to a different inventory and thus you easily implement policies like:
+
+> when a new version of the artifacts is built, deploy it first on staging.
+> Then, when it works on staging _promote_ that artifact version to production.
+
+
+All you need to do is to change the contents of the `inventory/stag.json` file to point to the next candidate version that you want to try out in staging and then once you detect that it's working fine (out of scope of this document), you just update the `inventory/prod.json` file to point to that same version.
+The file can be either updated manually or you can have an automation like [Mend Renovatebot](https://www.mend.io/renovate/) do it for you.
+
+Again, a pretty picture for art lovers among you:
+
+```mermaid
+%%{init: { 'theme': 'default', 'flowchart': {'curve': 'linear', 'rankSpacing': 120, 'nodeSpacing': 20} } }%%
+graph RL;
+    aws(aws)
+    prod-artifact(v1)
+    prod(prod)
+    app(app)
+    stag(stag)
+    stag-artifact(v2)
+    gcp(gcp)
+
+    aws
+    prod
+    app
+    stag
+    stag-artifact
+    gcp
+ 
+    prod-aws-us-east-1(us-east-1)-->prod-aws-common
+    prod-aws-us-west-1(us-west-1)-->prod-aws-common
+    prod-aws-common(common)-->aws & prod
+
+    prod-gcp-us-east-1(us-east-1)-->prod-gcp-common
+    prod-gcp-us-west-1(us-west-1)-->prod-gcp-common
+    prod-gcp-common(common)-->gcp & prod
+
+    stag-aws-us-east-1(us-east-1)-->stag-aws-common
+    stag-aws-us-west-1(us-west-1)-->stag-aws-common
+    stag-aws-common(common)-->aws & stag
+
+    stag-gcp-us-east-1(us-east-1)-->stag-gcp-common
+    stag-gcp-us-west-1(us-west-1)-->stag-gcp-common
+    stag-gcp-common(common)-->gcp & stag
+
+    prod-->prod-artifact
+
+    prod-->app
+    stag-->app
+
+    stag-->stag-artifact
+
+    style prod-artifact fill:#8f8
+    style stag-artifact fill:#8f8
+```
+
+The "production" template thus imports both the "app" template and the "v1" inventory JSON file. The "production" template overlays the "app" template with the "v1" inventory config, effectively causing every deployment that imports the staging template into a deployment that runs "v1" of the software.
+
+
+## Problem
+
+The main problem with this setup is that while each "phase" can have some specific overlays, and we make use of that feature to use a different version of the main application binaries in staging and in production, we have only one version of the main templates.
+
+> Any change to the templates applies immediately on every target cluster.
```

Comment:
```suggestion
> [!IMPORTANT]  
> Any change to the app templates applies immediately on every target cluster.
```

I think your intent was to call attention to this as a sort of "problem statement." This is Github-specific markdown, but would help with the callout. There are also "NOTE" and "WARNING" options.

![image](https://github.com/kubecfg/kubecfg/assets/2457559/ae6494db-dc67-4c7b-890b-e81fb5a729e4)


I've also clarified the templates to which you're referring.